### ISSUE TITLE
When editing the CC # on a recurring contribution, donor receives error message 

### ISSUE THREAD

**Issue by IowaBoy1** [iATSPayments/com.iatspayments.civicrm#252]
Drupal 7.59
CiviCRM 5.3.1
iATS 1.6.1

We have a donor that needs to change his credit card number. He gets an error message when he attempts to change his card.
![image](https://user-images.githubusercontent.com/28285267/43116204-600a3c52-8ebb-11e8-80c9-f8f8b5e3f3d7.png)

**Comment by adixon** [comment]
Can I confirm that you've enabled the self-service feature in the iATS configuration? Can you tell me the url (masking the id numbers) that you are using? Is the donor logged in? Is this a new issue after the 5.3.1 update?

**Comment by KarinG** [comment]
Last time I tested this was on Jun. 30 answering this SE question -> it has all the details I could think off at that time (and I confirmed it was working then): https://civicrm.stackexchange.com/questions/25446/only-logged-in-users-can-change-billing-and-recurring-amount
 
If you're positive you're doing all that - let me know along with whether or not your donor is logged out/in - and then I'll try reproduce on current 5.3.1

**Comment by IowaBoy1** [comment]
We do have self-service enabled.
Yes the donor is logged in. The instruction I gave him were to go to 
1. ourwebsite.org/login That page has a login block. 
2. Once logged in, he gets his Civi Donor Dashboard. Scroll down to the recurring contributions and click Change Billing Details of the one he wants to edit.
3. Edit and save the pop-up window
![screenshot 2018-07-24 16 45 46](https://user-images.githubusercontent.com/28285267/43171957-9b8af7a2-8f61-11e8-92bd-18eb170a9417.png)

**Comment by KarinG** [comment]
From the View Contact Dashboard -> the URL to update looks like:
civicrm/contribute/updatebilling?reset=1&crid=395&cid=2&context=dashboard

Got it -> same error:
![image](https://user-images.githubusercontent.com/5340555/43175358-e05340e6-8f7b-11e8-95bb-2d673ced1cb5.png)

Same error from within:
civicrm/contribute/updatebilling?reset=1&crid=395&cid=2&context=contribution

The admin pathway to update someone's card details: URLs like:
civicrm/contact/edit/iatscustomerlink?reset=1&cid=2&customerCode=A23190190&paymentProcessorId=33&is_test=0

Are working fine

Do you remember when this last worked for you? Or is this self-service something you recently started for your Donors?

Error - (install Alan's log Viewer -> handy Administer -> Administration Console -> View log):

```
[info] $backTrace = #0 /var/www/civicrm/5/CRM/Core/Error.php(381): CRM_Core_Error::backtrace("backTrace", TRUE)
#1 /var/www/civicrm/5/CRM/Contribute/Form/UpdateBilling.php(88): CRM_Core_Error::fatal("Required information missing.")
#2 /var/www/civicrm/5/CRM/Core/Form.php(562): CRM_Contribute_Form_UpdateBilling->preProcess()
#3 /var/www/civicrm/5/CRM/Core/QuickForm/Action/Display.php(92): CRM_Core_Form->buildForm()
#4 /var/www/civicrm/5/packages/HTML/QuickForm/Controller.php(203): CRM_Core_QuickForm_Action_Display->perform(Object(CRM_Contribute_Form_UpdateBilling), "display")
#5 /var/www/civicrm/5/packages/HTML/QuickForm/Page.php(103): HTML_QuickForm_Controller->handle(Object(CRM_Contribute_Form_UpdateBilling), "display")
#6 /var/www/civicrm/5/CRM/Core/Controller.php(351): HTML_QuickForm_Page->handle("display")
#7 /var/www/civicrm/5/CRM/Utils/Wrapper.php(113): CRM_Core_Controller->run()
#8 /var/www/civicrm/5/CRM/Core/Invoke.php(283): CRM_Utils_Wrapper->run("CRM_Contribute_Form_UpdateBilling", "Update Billing Details", NULL)
#9 /var/www/civicrm/5/CRM/Core/Invoke.php(84): CRM_Core_Invoke::runItem((Array:14))
#10 /var/www/civicrm/5/CRM/Core/Invoke.php(52): CRM_Core_Invoke::_invoke((Array:3))
#11 /var/www/civicrm/5/drupal/civicrm.module(445): CRM_Core_Invoke::invoke((Array:3))
#12 [internal function](): civicrm_invoke("contribute", "updatebilling")
#13 /var/www/drupal/git-7/includes/menu.inc(527): call_user_func_array("civicrm_invoke", (Array:2))
#14 /var/www/drupal/git-7/index.php(21): menu_execute_active_handler()
#15 {main}
```

So it looks like CRM_Core_Error::fatal("Required information missing.") is the key thing here.

**Comment by IowaBoy1** [comment]
We have been using the self-service for a long time. We started with Civi about 2+ years and I think we used it when we initially set up.  I do not know when it failed. Friday July 20th, we updated CiviCore. The donor notified me on Sat that he could not edit this. So I assume it started with this update but...

**Comment by KarinG** [comment]
Hi Carl - what version of CiviCRM were you on prior to Friday's upgrade?

**Comment by adixon** [comment]
Thanks for this - I didn't even know about this method of enabling self-service cc updates to credit card info! I've been able to reproduce and assume it's a 5.x compatibility issue, we've found a few of them. Will get back to your shortly and hope to include in the next update.

**Comment by KarinG** [comment]
@adixon - not sure this is related but see: https://lab.civicrm.org/dev/core/issues/289 - other Edit screens having throuble post 5.3.1 security update

**Comment by IowaBoy1** [comment]
Is there any progress being made on a fix for this? I am getting regular requests from donors to be able to edit their recurring donation bank/card info.

**Comment by adixon** [comment]
I've just finished tracking this down - it was a bit complicated because this functionality was contributed to us, and depended on some core code that changed in some subtle ways (i.e. the recurring contribution id stopped getting passed into the iATS code that updated the credit card information). I'll put in a simple fix that will require a small core patch, longer term we'll need a better way of implementing this function.

**Comment by adixon** [comment]
I'll be releasing 1.6.2 shortly - but note that you'll need to implement that core patch for this to work.

**Comment by IowaBoy1** [comment]
I have updated Civi to 5.4.1 and applied the patch above. I had a recurring contribution setup. After the updates I signed into my donor dashboard to edit the CC info. I changed to my other credit card and it showed a popup that gave me a Authorization code and OK, I clicked OK and it gave another popup that said that the Recurring contribution billing info was successfully updated. I clicked OK and it gave a 3rd popup that said it could not connect to the network server.

I do not have any errors in Allan's logging. It has 3 entries, Contribution update success, receipt sent, and  Database Updated. However, in the Drupal logs I do have warnings and such. 

access denied | 09/07/2018 - 21:30 | civicrm | Carl_Donor |  
php | 09/07/2018 - 21:30 | Strict warning: Only variables should be assigned by... | Carl_Donor |  
php | 09/07/2018 - 21:30 | Warning: Invalid argument supplied for foreach() in... | Carl_Donor |  
php | 09/07/2018 - 21:30 | Notice: Undefined property: CRM_Contribute_Form... | Carl_Donor

Details of Notice Undefined Property
Notice: Undefined property: CRM_Contribute_Form_UpdateBilling::$_fields in CRM_Contribute_Form_UpdateBilling->postProcess() (line 228 of /sites/all/modules/civicrm/CRM/Contribute/Form/UpdateBilling.php).

Details of Warning: Invalid argument
Warning: Invalid argument supplied for foreach() in CRM_Contribute_Form_UpdateBilling->postProcess() (line 228 of /sites/all/modules/civicrm/CRM/Contribute/Form/UpdateBilling.php).

Details of Strict warning: Only variables
Strict warning: Only variables should be assigned by reference in CRM_Contribute_Form_UpdateBilling->postProcess() (line 257 of /sites/all/modules/civicrm/CRM/Contribute/Form/UpdateBilling.php).

Details of Access Denied
There really is not much info there. message just says civicrm

Is all this anything to be concerned with if it worked?

**Comment by adixon** [comment]
It would seem like in the civi core postProcessing of the billing update is the issue. If it worked, then I suspect there's just some kind of weird core code issue or it's related to the extra field my patch added.

Tell me if you can reproduce it reliably.

**Comment by IowaBoy1** [comment]
I just changed my billing info 3 xs and got the same results as above all 3 times

**Comment by IowaBoy1** [comment]
We are still getting the error message when the card info is edited. It looks to donors as if it failed but it did actually work.

"I have updated Civi to 5.4.1 and applied the patch above. I had a recurring contribution setup. After the updates I signed into my donor dashboard to edit the CC info. I changed to my other credit card and it showed a popup that gave me a Authorization code and OK, I clicked OK and it gave another popup that said that the Recurring contribution billing info was successfully updated. **I clicked OK and it gave a 3rd popup that said it could not connect to the network server.**"

Any ideas?
We are now on Drupal 7.61 & Civi 5.4.1
